[FEAT] Frogs

tiffanydys opened the pull request:
# Description
Frogs is a community driven feature that is made by the community team (Project Dignity : @tiffanydys, @DonatasDee, @VadimasVP, @Paluras)

Discussion #1217 

## Type of change
- [x] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?
`yarn test`
<img width="400" alt="Screen Shot 2022-08-18 at 4 21 41 AM" src="https://user-images.githubusercontent.com/18549210/185235580-c32bb47d-7fbf-4602-8db6-d7c2169b5689.png">

`yarn dev`
- to test in localhost please DM me your wallet address so I can whitelist you and send you 300 SFL for test minting
- make sure to update your `.env` file from `.env.sample`
- to everyone who wants to view test dummy data of your minted frogs you can go to `src/features/game/actions/loadFrogs.ts` and uncomment line number 15 and comment out line 16:

<img width="201" alt="Screen Shot 2022-08-18 at 4 32 08 AM" src="https://user-images.githubusercontent.com/18549210/185237456-2b0807bf-90b1-4e84-a4ae-d476da254ac7.png">

### Modals
#### Approve Spending Modal
<img width="400" src="https://user-images.githubusercontent.com/18549210/186382267-d72b7564-4f13-4be7-ad4e-85a3e057498f.png">

#### Mint Modal
<img width="400" src="https://user-images.githubusercontent.com/18549210/186384723-4cb98191-cf36-4555-b9fa-74d6de0657c5.png">

#### Minted Modal
<img width="400" src="https://user-images.githubusercontent.com/18549210/186407828-f27ef4b5-0c06-49a3-afd7-afeac37d22cf.png">

#### Blacklisted Modal (After minting 1 Frog)
<img width="400" src="https://user-images.githubusercontent.com/18549210/186407742-5f6c8d72-1b59-4e4b-ba69-052eeda106c8.png">

### Not Whitelisted
<img width="400" src="https://user-images.githubusercontent.com/18549210/186405509-97defd57-d5a4-45aa-81d0-75020c27e7f4.png">

### Event Ended
<img width="400" src="https://user-images.githubusercontent.com/18549210/186405731-be4b8cf6-7b59-4a72-9abd-6e87a0ac9410.png">

# Checklist:

- [x] Title of the PR is relevant and is prefixed with [FEAT], [CHORE] or [FIX]
- [x] Screenshot if it includes any UI changes
- [x] I have read the contributing guidelines and agree to the T&Cs
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

j923sneaks commented on src/features/goblins/merchant/lib/frogMachine.ts:
```diff
@@ -0,0 +1,228 @@
+
+            return { _approve };
+          },
+          onDone: [
```
can omit `[]`

j923sneaks commented on src/features/goblins/merchant/lib/frogMachine.ts:
```diff
@@ -0,0 +1,228 @@
+
+            return { mint };
+          },
+          onDone: [
```
based on `mintFrog()`, `verified` always return `true` so i think there's no need for `cond`. also can omit `[]` here in `onDone`

tiffanydys replied:
You're right, will refactor this thanks!

j923sneaks commented on src/lib/blockchain/metamask.ts:
```diff
@@ -328,6 +331,11 @@ export class Metamask {
   }
 
+  public getFrog() {
+    console.log("frog");
```
cleanup

j923sneaks reviewed (commented):
Hi @tiffanydys , Awesome job! Will run on my local but here are some initial comments :)

tiffanydys commented on src/features/goblins/merchant/MerchantModal.tsx:
```diff
@@ -0,0 +1,202 @@
+              </Button>
+              <Button
+                className="text-xs mt-2"
+                onClick={() => {
```
Oh right I forgot to update that LOL thanks

tiffanydys replied:
Button links added ✌🏻

j923sneaks commented on .env.sample:
```diff
@@ -17,6 +17,7 @@ VITE_DISCORD_REDIRECT=http://localhost:3000
 VITE_RECAPTCHA_SITEKEY=6Lfqm6MeAAAAAFS5a0vwAfTGUwnlNoHziyIlOl1s
 VITE_TRADER_CONTRACT=0xbE005Fd6328960413f2627Ebb489B09507B2AEdF
 VITE_TREASURY_ADDRESS=0x1DB57407EE80709D4d862fe81399FBB35B8B9586
+VITE_FROG_CONTRACT=0x52A5ddFF59E0A27a1cBA75c8045A434b7D296A6c
```
does not match with the one in testnet.yaml

tiffanydys replied:
Updated this one, thanks!

veenhdek commented:
Tested just now after getting it to launch (.env issues i should have thought of) went pretty smooth I put a full report in discord 😄 https://discord.com/channels/880987707214544966/905567246452129873/1008064550924464149 🚀

adamhannigan commented on src/features/goblins/merchant/MerchantModal.tsx:
```diff
@@ -0,0 +1,271 @@
+import { ErrorCode } from "lib/errors";
+
+// images
+import frog_unrevealed from "../../../assets/nfts/frogs/frog_unrevealed.gif";
```
We can use absolute paths for `assets`.

E.g. `assets/nfts/frogs/frog_unrevealed.gif`

tiffanydys replied:
Used absolute paths for this file and other files as well 😄

adamhannigan commented on src/lib/blockchain/Token.ts:
```diff
@@ -77,4 +80,34 @@ export class Token {
+    return approve;
+  }
+
+  public async isTokenApprovedForFrogMint(attempts = 0): Promise<boolean> {
```
Could we make this more reusable for other community projects and let the caller pass in the contract

I.e. `isTokenApprovedForContract(address: string, attempts = 0):`

tiffanydys replied:
Good catch! I'll update this as well. ❤️

adamhannigan commented on src/lib/blockchain/metamask.ts:
```diff
@@ -66,6 +68,7 @@ export class Metamask {
         this.account as string
       );
       this.trader = new Trader(this.web3 as Web3, this.account as string);
+      this.frog = new Frog(this.web3 as Web3, this.account as string);
```
Since the core team does not have control over this contract we will need an extra `try / catch` clause.

This will be helpful if a community contract decided to destroy the contract (which could cause a whijte screen in the game 😓 )

I.e

```
try {
 this.frog = new Frog(this.web3 as Web3, this.account as string);
} catch (e) {
console.error("Unable to initialise frog contract", e)
}
```

adamhannigan commented on src/features/game/actions/loadFrogs.ts:
```diff
@@ -0,0 +1,128 @@
+  headers.append("Content-Type", "application/json");
+  headers.append("Accept", "application/json");
+
+  const url = `${FROG_URL}/frogs/${request.frogId}`;
```
To follow ERC721 standards, we should fetch this URL directly from the smart contract instead of hard coding the `FROG_URL`.

For instance.

`const url = metamask.getFrog().getTokenUri(id)`

This would return the URL from the on-chain data

tiffanydys replied:
Hi adam!

The frog URI (metadata URL) is different from the `FROG_URL` (which is the API) I think I would have named it better as FROG_API, I think it would be saver adding it to the .env file rather than adding it to the contract. What do you think? 😅

tiffanydys replied:
Done! ✌🏻😄

adamhannigan reviewed (commented):
Some work in progress feedback 🤗 

Just waiting for a whitelisting (DMed) you and then I can review the UI.

adamhannigan commented:
@tiffanydys and developers 👋 

One of the aims of introducing community content is to give ownership and scale features quickly. A big part of this is allowing teams to iterate and work at their pace without too much friction - This means there is a risk of the core app breaking due to a bug.

We need to look at architecture/separation of code from the core domain/repo. This will help both sides scale without impacting/breaking the other sides work. Right now where my head is at is creating a dedicated "Community" area that would have it's own map and dedicated "/community" URL. You would still access this through the Merchant on Goblin Land.

We can manage the initial setup of this area with a unique space, but we will need to start moving the code in this PR into a new area `src/features/community`. The community area can connect directly with a user's wallet without needing to go through the typical core authentication flow (this will actually simplify a lot of the features). In terms of refactoring, for example we could move the FrogContract out of `metamask.ts` and into a new `src/features/community/lib/communityContracts.ts` file.

It is a bit of work that I wish I anticipated from the start. Now that I see the code in action, it feels like the right forward to separate these concerns and let each team move at their own pace without fear of breaking one another's code 😅 

Let me create some assets/diagram today to help explain this flow of a "community" area 😄

spencerdezartsmith reviewed (commented):
Great job so far! Some of the things that I noticed..

*Functionality*
- Once I approved the SFL I was take back to the "approve" modal. I needed to close this and reopen to be able to mint. It would be better to be taken to the "mint" modal automatically after the successful approve tx.

*UI*
- I think we need to do a better job communicating that this is a community feature and what that means. We have spoken about it and Adam is going to reach out with some of the ideas we had for this around creating a seperate community space. 
- I also this it wasn't very clear that I needed to have SFL in my metamask wallet and not my farm wallet. This is a tough message to convey. I did see it in the warning message with a link but I don't think people will click on that tbh. You all might need to think about how to make this message clearer outside of the warning message.
- I also would like to move away from the `error` red colour in the warning messages. We have started using a more traditional warning colour of `orange` which is less threatening 😅 . 
- The style of some of the modals flip our style upside down 🙃 . The whitelisted ones have the text on top and then the image. We typically have it in reverse so I would like to be consistent with that.
- The button should be full width for all of the modals so including the `Mint a frog` one imo. 
- The padding in the modals should be consistent. There is a lot of padding in the minted modals but then very little in the Blacklisted modal . 
- The close button on the modals appears to be `block` and taking up the whole row which looks a little off. 

We are working on setting up some generic modal components to align these styles better to make it easier for devs to not have to "redesign" with every feature. Until we have that we want to try our best to be some what consistent in the design. 

This is coming along nicely though. Can't wait for the release 🐸

adamhannigan commented:
Here is the PR that implements a community style area.

Ideally we want all the new code/actions etc. to live inside of the `features/community` directory.

I have also introduced a community machine that loads the balance from the **personal wallet** instead of the farm wallet. This means that the HUD will show the amount that is in the personal wallet, which is more ideal than the Goblin HUD which shows the balance from the farm's wallet

https://github.com/sunflower-land/sunflower-land/pull/1348

AzaroQL commented:
I would propose to make a route to community place from the farm, otherwise it requires two redirects to get there(farm->goblin->community).
How about this place with an arrow and a merchant, same way as for goblin?
<img src="https://user-images.githubusercontent.com/9656961/185723555-ad5f5f24-0dc4-4a40-a822-9020e274cc69.png" height="200">

tiffanydys commented:
Hey guys thank you so much for the reviews! ❤️

@adamhannigan Thank you for initiating the community garden 😊, I've successfully translated the files to fit the `/community` directory.

@spencerdezartsmith I agree, the modals do have different paddings 😅 I've followed your comments and I've updated the images of the modals on this PR. Please have a look and tell me if there are any more adjustments. ❤️

@AzaroQL Thanks for the inspiration on this. ❤️  I've applied it and moved the yellow book a bit to the bottom. 👀
<img width="450px" src="https://user-images.githubusercontent.com/18549210/186426718-e648edce-30f2-4a56-a996-80c08fbd9696.png" />

**Modal to Community Garden:**
<img width="450px" src="https://user-images.githubusercontent.com/18549210/186427295-1145686a-40ca-41a6-9ec6-5bf4e479177e.png" />

AzaroQL commented on src/features/community/CommunityGardenEntry.tsx:
```diff
@@ -0,0 +1,71 @@
+    merchantAudio.play();
+    setShowModal(true);
+    //Checks if merchantAudio is playing, if false, plays the sound
+    if (!merchantAudio.playing()) {
```
There are cases when first merchantAudio.play() wont work?

AzaroQL replied:
Tested it and its actually line 15 that should be removed. IsPlaying check is here to prevent overlaping sound when you open and close modal too fast.

tiffanydys replied:
removed extra `merchantAudio.play()` thanks!

AzaroQL commented on src/features/community/CommunityGardenEntry.tsx:
```diff
@@ -0,0 +1,71 @@
+import React, { useState } from "react";
```
Shouldnt CommunityGardenEntry and CommunityGardenModal files be placed in src\features\farming\town\components, in the same place as GoblinVillageEntry and GoblinVillageModal?
Those components are displayed on farming screen, so to prevent confusion its better to place them there.

Maybe even create subfolders communityGarden and goblinVillage.

AzaroQL replied:
Or other way around, goblin files should be moved out. @adamhannigan whats your opinion on that?

tiffanydys replied:
refactored and moved files to `src/features/farming/town/components` thanks!

AzaroQL commented on src/components/ui/FrogNft.tsx:
```diff
@@ -0,0 +1,55 @@
+  const [frogData, setFrogData] = useState([{ pixel_image: "" }]);
+
+  const open = () => {
+    setShowModal(true);
```
There is no modal to open?

tiffanydys replied:
Not yet there isn't 👀 (there will be in my new update)

tiffanydys replied:
On second thought, it doesn't need a modal so I've removed it 😅. Thanks!

AzaroQL commented on src/features/community/merchant/MerchantModal.tsx:
```diff
@@ -0,0 +1,287 @@
+              </p>
+              <p className="text-xxs mb-4\2 mt-2 text-center">
+                initiated by{" "}
+                <a href={`${projectDignity}/project-dignity/members`}>
```
This link should be opened in new tab.

tiffanydys replied:
applied thanks!

AzaroQL commented on src/features/community/CommunityGardenModal.tsx:
```diff
@@ -0,0 +1,51 @@
+        </div>
+
+        <div className="flex">
+          <Button className="mr-1" onClick={handleClose}>
```
This button doesnt work

tiffanydys replied:
Hey @AzaroQL can you help me on this? I can't seem to make it work whatever I try lmao

AzaroQL replied:
You need to move modal out of div that has onClick event.

tiffanydys replied:
Solved, thanks! 😄❤️

AzaroQL commented on src/features/community/merchant/Merchant.tsx:
```diff
@@ -0,0 +1,55 @@
+  const [isOpen, setIsOpen] = React.useState(false);
+
+  const openMerchant = () => {
+    merchantAudio.play();
```
Check my other comment regarding overlapping audio.

tiffanydys replied:
removed thanks!

AzaroQL commented on src/features/community/components/BackButton.tsx:
```diff
@@ -0,0 +1,23 @@
+export const BackButton: React.FC<Props> = ({ farmId }) => {
+  const goBackToFarm = () => {
+    console.log(farmId);
+    window.location.href = `${window.location.pathname}#/farm/${farmId}`;
```
Check src\features\goblins\components\TownEntryModal.tsx its better to use that approach for back button. You dont need to pass farmId to that component and it seems faster to load, but that can be just me.

tiffanydys replied:
Thanks for this, I never thought of using navigation for it, much cleaner and yes much faster than passing the prop ✌🏻😄

adamhannigan commented on src/features/community/merchant/actions/loadFrogs.ts:
```diff
@@ -0,0 +1,127 @@
+};
+
+export async function loadFrogs() {
+  let frogIds = await communityContracts.getFrog().getFrogIds();
```
Could we wrap all of this logic in a try/catch. Since this will be used in the main application, we need to support cases where the contracts/APIs could be down.

For example.

```
try {
 let frogIds = await communityContracts.getFrog().getFrogIds();
  ...
} catch {
   return null
}
```

tiffanydys replied:
added a try catch on load frogs, thanks ✌🏻

adamhannigan commented on src/features/community/merchant/actions/loadFrogs.ts:
```diff
@@ -0,0 +1,127 @@
+  headers.append("Content-Type", "application/json");
+  headers.append("Accept", "application/json");
+
+  const frogBaseUri = await communityContracts.getFrog().getBaseUri();
```
<img width="348" alt="Screen Shot 2022-08-29 at 8 48 26 am" src="https://user-images.githubusercontent.com/11745561/187097729-238f1f09-64ad-4d09-9804-423fdec8a4ad.png">

Calling the `tokenUri` directly on an ERC721 contract will return the full version of the URL. This means you don't need to do concatenation below.

tiffanydys replied:
I forgot there's a tokenURI function 😅 thanks adam

adamhannigan reviewed (commented):
Nice work @tiffanydys 👏 

There are a few comments by @AzaroQL that need to be addressed. I have included a suggestion for using the `tokenURI` and also adding a `try/catch` around some of the core logic since it will live inside of the game.

Once these are satisfied, let's move forward and launch the frogs 🚀 

Before adding the mainnet address, could you please verify the source code on PolygonScan - if you visit the contract address you will be able to view the contract and upload the source code 😄

adamhannigan commented:
<img width="158" alt="Screen Shot 2022-08-29 at 8 56 34 am" src="https://user-images.githubusercontent.com/11745561/187097958-d9dbde5c-c634-4865-9bbd-e0e036510828.png">

Btw, very happy with my frog! ❤️

AzaroQL commented:
@tiffanydys Due to recent changes in main branch you will need to apply same fix for community garden.
https://github.com/sunflower-land/sunflower-land/pull/1392/files

adamhannigan commented on .github/workflows/testnet.yml:
```diff
@@ -40,6 +40,7 @@ jobs:
           VITE_PAIR_CONTRACT: "0xaE42EA9864C36820627068a11959915f817ce809"
           VITE_MOM_CONTRACT: "0x2cFdb89954991b69D6D4B2ec8b70f3b995F4E8Ec"
           VITE_TRADER_CONTRACT: "0xbE005Fd6328960413f2627Ebb489B09507B2AEdF"
+          VITE_FROG_CONTRACT: "0x52A5ddFF59E0A27a1cBA75c8045A434b7D296A6c"
```
Is the testnet address meant to be the same as the mainnet address?

tiffanydys replied:
Yes, they're the same contract:

Testnet address:
https://mumbai.polygonscan.com/address/0x52a5ddff59e0a27a1cba75c8045a434b7d296a6c

Mainnet address:
https://polygonscan.com/address/0x52a5ddff59e0a27a1cba75c8045a434b7d296a6c

I didn't change anything from the code on the testnet, I think that's why it generated the same address 😅

adamhannigan reviewed (approved):
@tiffanydys LGTM. There are a few small tweaks but I feel releasing and getting feedback will be more important for the project at this stage. 

What timeline were you looking at to launch this feature?

tiffanydys commented:
> @tiffanydys LGTM. There are a few small tweaks but I feel releasing and getting feedback will be more important for the project at this stage. 
> 
> 
> 
> What timeline were you looking at to launch this feature?

hey @adamhannigan this code is up for review 🤗 though i'm not sure as to why the community garden doesn't scroll to the sides when viewed on mobile, any ideas why it doesn't work? 😅

adamhannigan reviewed (approved):
Great work @tiffanydys 👏 

I found the issue with the map scrolling and made a few adjustments. It required a new container and some positioning. I have pushed this changes in

You may want to check the new positioning of the seal + aquatics to ensure they are now in the right spot. Otherwise happy to merge this in.

Do you have an exact timeline for when the feature will go live and when the whitelisting will begin?

Once you share those dates we can merge this in and align on announcements for the community 😄

DonatasDee commented:
> Great work @tiffanydys 👏
> 
> I found the issue with the map scrolling and made a few adjustments. It required a new container and some positioning. I have pushed this changes in
> 
> You may want to check the new positioning of the seal + aquatics to ensure they are now in the right spot. Otherwise happy to merge this in.
> 
> Do you have an exact timeline for when the feature will go live and when the whitelisting will begin?
> 
> Once you share those dates we can merge this in and align on announcements for the community 😄

Our whitelist is almost filled up and we were waiting to finish it once we knew when the code is ready for merge, we'd be glad to allow the whitelists to mint this Friday and then do public mint the next day/s. If that's an okay date then we can start organizing around that. ❤️
